File: backend/app/api/account_bots.py
"""账号绑定普通 Bot 联动 API。"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from sqlalchemy import select

from ..db.models.account_bot import ACCOUNT_BOT_STATUS_DISABLED, AccountBot
from ..deps import CurrentUser, DBSession
from ..schemas.account_bot import (
    AccountBotConfigResponse,
    AccountBotConfigUpdate,
    AccountBotRemotePluginPolicy,
    AccountBotRuntimeResponse,
    AccountBotTestRequest,
    AccountBotTestResponse,
    AccountBotUserCreate,
    AccountBotUserResponse,
    AccountBotUserUpdate,
)
from ..services import account_bot_runtime, account_bot_service, audit

router = APIRouter(prefix="/api/accounts", tags=["account-bots"])


@router.get("/{aid}/bot", response_model=AccountBotConfigResponse)
async def get_account_bot(
    aid: int,
    db: DBSession,
    _user: CurrentUser,
) -> AccountBotConfigResponse:
    """读取该账号 Bot 配置；不返回 token 明文。"""

    await account_bot_service.ensure_account(db, aid)
    row = (
        await db.execute(select(AccountBot).where(AccountBot.account_id == aid))
    ).scalar_one_or_none()
    if row is None:
        return AccountBotConfigResponse(
            account_id=aid,
            enabled=False,
            status=ACCOUNT_BOT_STATUS_DISABLED,
            has_token=False,
            remote_plugin_policy=AccountBotRemotePluginPolicy(),
        )
    return account_bot_service.config_to_response(row)


@router.put("/{aid}/bot", response_model=AccountBotConfigResponse)
async def update_account_bot(
    aid: int,
    payload: AccountBotConfigUpdate,
    db: DBSession,
    user: CurrentUser,
) -> AccountBotConfigResponse:
    """保存该账号 Bot token/启停配置，并同步 polling runtime。"""

    row = await account_bot_service.update_bot_config(db, aid, payload)
    await audit.write(
        db,
        user.id,
        "account_bot.update",
        target=f"account:{aid}/bot",
        detail={
            "enabled": payload.enabled,
            "token_changed": bool(payload.bot_token or payload.clear_token),
            "remote_plugin_policy_changed": payload.remote_plugin_policy is not None,
        },
    )
    await db.commit()
    await db.refresh(row)
    await account_bot_runtime.sync_account_bot(aid)
    return account_bot_service.config_to_response(row)


@router.post("/{aid}/bot/test", response_model=AccountBotTestResponse)
async def test_account_bot(
    aid: int,
    payload: AccountBotTestRequest,
    db: DBSession,
    user: CurrentUser,
) -> AccountBotTestResponse:
    """向已授权且有 last_chat_id 的通知用户发送测试消息。"""

    row = await account_bot_service.get_bot_config(db, aid, create=False)
    token = account_bot_service.decrypt_bot_token(row)
    users = await account_bot_service.list_bot_users(db, aid)
    targets = [
        u for u in users
        if u.enabled and u.notify_enabled and u.last_chat_id is not None
    ]
    if not targets:
        raise HTTPException(
            400,
            detail={
                "code": "ACCOUNT_BOT_NO_TARGET",
                "message": "没有可发送的授权用户。请先让授权用户给这个 Bot 发送 /start。",
            },
        )
    text = payload.text or "✅ TelePilot 账号 Bot 测试消息发送成功。"
    sent = 0
    last_error = None
    for target in targets:
        try:
            await account_bot_service.send_message(token, int(target.last_chat_id), text)
            sent += 1
        except Exception as exc:  # noqa: BLE001
            last_error = account_bot_service.sanitize_bot_error(exc, token=token)
    await audit.write(
        db,
        user.id,
        "account_bot.test",
        target=f"account:{aid}/bot",
        detail={"sent": sent},
    )
    await db.commit()
    if sent <= 0:
        raise HTTPException(
            502,
            detail={
                "code": "ACCOUNT_BOT_TEST_FAILED",
                "message": last_error or "测试发送失败",
            },
        )
    return AccountBotTestResponse(ok=True, sent=sent, message="测试消息已发送")


@router.post("/{aid}/bot/restart-runtime", response_model=AccountBotRuntimeResponse)
async def restart_account_bot_runtime(
    aid: int,
    db: DBSession,
    user: CurrentUser,
) -> AccountBotRuntimeResponse:
    """重启该账号 Bot polling task。"""

    await account_bot_service.ensure_account(db, aid)
    await audit.write(
        db,
        user.id,
        "account_bot.restart_runtime",
        target=f"account:{aid}/bot",
    )
    await db.commit()
    await account_bot_runtime.restart_account_bot(aid)
    return AccountBotRuntimeResponse(ok=True, message="已重启 Bot polling runtime")


@router.get("/{aid}/bot/users", response_model=list[AccountBotUserResponse])
async def list_account_bot_users(
    aid: int,
    db: DBSession,
    _user: CurrentUser,
) -> list[AccountBotUserResponse]:
    rows = await account_bot_service.list_bot_users(db, aid)
    return [AccountBotUserResponse.model_validate(r) for r in rows]


@router.post("/{aid}/bot/users", response_model=AccountBotUserResponse, status_code=201)
async def create_account_bot_user(
    aid: int,
    payload: AccountBotUserCreate,
    db: DBSession,
    user: CurrentUser,
) -> AccountBotUserResponse:
    row = await account_bot_service.create_bot_user(db, aid, payload)
    await audit.write(
        db,
        user.id,
        "account_bot_user.create",
        target=f"account:{aid}/bot_user:{row.tg_user_id}",
        detail={"role": row.role, "notify_enabled": row.notify_enabled},
    )
    await db.commit()
    await db.refresh(row)
    return AccountBotUserResponse.model_validate(row)


@router.patch("/{aid}/bot/users/{uid}", response_model=AccountBotUserResponse)
async def update_account_bot_user(
    aid: int,
    uid: int,
    payload: AccountBotUserUpdate,
    db: DBSession,
    user: CurrentUser,
) -> AccountBotUserResponse:
    row = await account_bot_service.update_bot_user(db, aid, uid, payload)
    await audit.write(
        db,
        user.id,
        "account_bot_user.update",
        target=f"account:{aid}/bot_user:{row.tg_user_id}",
        detail=payload.model_dump(exclude_unset=True),
    )
    await db.commit()
    await db.refresh(row)
    return AccountBotUserResponse.model_validate(row)


@router.delete("/{aid}/bot/users/{uid}", response_model=dict)
async def delete_account_bot_user(
    aid: int,
    uid: int,
    db: DBSession,
    user: CurrentUser,
) -> dict[str, bool]:
    row = await account_bot_service.get_bot_user(db, aid, uid)
    tg_user_id = row.tg_user_id
    await account_bot_service.delete_bot_user(db, aid, uid)
    await audit.write(
        db,
        user.id,
        "account_bot_user.delete",
        target=f"account:{aid}/bot_user:{tg_user_id}",
    )
    await db.commit()
    return {"ok": True}

File: frontend/src/api/accountBots.ts
import { api } from "@/lib/api";
import type {
  AccountBotConfig,
  AccountBotConfigUpdate,
  AccountBotInteractionConfig,
  AccountBotInteractionConfigUpdate,
  AccountBotTestRequest,
  AccountBotTestResponse,
  AccountBotUser,
  AccountBotUserCreate,
  AccountBotUserUpdate,
} from "@/api/types";

export async function getAccountBot(aid: number): Promise<AccountBotConfig> {
  const { data } = await api.get<AccountBotConfig>(`/api/accounts/${aid}/bot`);
  return data;
}

export async function updateAccountBot(
  aid: number,
  payload: AccountBotConfigUpdate,
): Promise<AccountBotConfig> {
  const { data } = await api.put<AccountBotConfig>(`/api/accounts/${aid}/bot`, payload);
  return data;
}

export async function testAccountBot(
  aid: number,
  payload?: AccountBotTestRequest,
): Promise<AccountBotTestResponse> {
  const { data } = await api.post<AccountBotTestResponse>(
    `/api/accounts/${aid}/bot/test`,
    payload ?? {},
  );
  return data;
}

export async function restartAccountBotRuntime(aid: number): Promise<void> {
  await api.post(`/api/accounts/${aid}/bot/restart-runtime`);
}

export async function getInteractionBotConfig(
  aid: number,
): Promise<AccountBotInteractionConfig> {
  const { data } = await api.get<AccountBotInteractionConfig>(
    `/api/accounts/${aid}/interaction-bot`,
  );
  return data;
}

export async function updateInteractionBotConfig(
  aid: number,
  payload: AccountBotInteractionConfigUpdate,
): Promise<AccountBotInteractionConfig> {
  const { data } = await api.put<AccountBotInteractionConfig>(
    `/api/accounts/${aid}/interaction-bot`,
    payload,
  );
  return data;
}

export const getTransferNoticeConfig = getInteractionBotConfig;
export const updateTransferNoticeConfig = updateInteractionBotConfig;

export async function listAccountBotUsers(aid: number): Promise<AccountBotUser[]> {
  const { data } = await api.get<AccountBotUser[]>(`/api/accounts/${aid}/bot/users`);
  return data;
}

export async function createAccountBotUser(
  aid: number,
  payload: AccountBotUserCreate,
): Promise<AccountBotUser> {
  const { data } = await api.post<AccountBotUser>(
    `/api/accounts/${aid}/bot/users`,
    payload,
  );
  return data;
}

export async function updateAccountBotUser(
  aid: number,
  uid: number,
  payload: AccountBotUserUpdate,
): Promise<AccountBotUser> {
  const { data } = await api.patch<AccountBotUser>(
    `/api/accounts/${aid}/bot/users/${uid}`,
    payload,
  );
  return data;
}

export async function deleteAccountBotUser(aid: number, uid: number): Promise<void> {
  await api.delete(`/api/accounts/${aid}/bot/users/${uid}`);
}
